### PR TITLE
mycroft: init at 0.9.19 [WIP]

### PR DESCRIPTION
**Pull request by peterhoeg**
###### Motivation for this change

Well, who doesn't want AI on your NixOS machine?

This is *VERY* much WIP but let's get it out there as soon as possible.

What works:

1. The services can be run manually or using the systemd services
2. Everything comes up (and doesn't crash) when run inside a VM using ```nixos-rebuild build-vm```.

But I don't *actually* know if it works, because there are a number of things that need to be fixed first:

1. We need to handle how to run a pulseaudio instance when it's not running inside the user session
2. All the python packages need to be separated out into their own packages
3. Tests need to be enabled for the various packages (I just disabled them if they were causing problems)
4. Mycroft itself has 2 failing tests - I haven't looked at all at why that isn't working
5. The skill script (msm) is borked (and not in the right place)
6. Need to figure out a way to have skills that are preloaded merged with skills dynamically installed - probably unionfs
7. Enclouse daemon isn't working

Cc: @dotlambda @bjornfor 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by oxij** [comment]
Can you, please, elaborate on PulseAudio issue?

**Comment by peterhoeg** [comment]
>   Can you, please, elaborate on PulseAudio issue?

Sure - it isn't actually so much of a pulseaudio issue as it's a "which run-mode do we use".

Mycroft needs to run as an unprivileged user and we can do this in 3 ways:

1. A normal script fired off by a user (this is what happens when you run the upstream-sanctioned start-mycroft.sh) which backgrounds a bunch of services. This is pretty ugly but allows mycroft to interact with the user's running pulseaudio instance.

2. A bunch of systemd user services which can also talk to the user's pulseaudio instance.

3. A bunch of systemd *system* services that require a system-wide pulseaudio instance.

So either mycroft should run in "appliance mode" if deployed to a Raspberry Pi or similar or in user-mode when used by a regular user. We then have to bring in pulseaudio in system mode if running as an appliance and probably fiddle that paths if running in user-mode.

File: nixos/modules/services/misc/mycroft.nix [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    hardware.pulseaudio.systemWide = lib.mkIf cfg.applianceMode true;
+
+    systemd = {
+      packages = with pkgs; [ mycroft ];
+      targets.mycroft.wantedBy = lib.mkIf cfg.applianceMode [ "multi-user.target" ];
```
Comment (dotlambda) [inline nixos/modules/services/misc/mycroft.nix]: If I understand it correctly, nothing is run if `services.mycroft.enable = true`. What's the purpose of having two "enable" options, then?

**Comment by peterhoeg** [comment]
>   true. What's the purpose of having two "enable" options, then?

It gives you access to the systemd units so you can do this:

``` shell
systemctl --user start mycroft.target
```

It seems a little heavy-handed to enable it for ALL users.

**Comment by matthiasbeyer** [comment]
Hi, I'm excited about this PR. I have some questions:

* What does the `mycroft-cli` actually do? Does it run a mycroft instance, so am I able to use this for a first test of mycroft?
* What STT engine is used by this package as default? I don't want any $bigCorp to listen, so does this use the mozilla speech recognition system on my own hardware without accessing the internet? If not, maybe this would be a nice idea as a package/service parameter/option.

**Comment by matthiasbeyer** [comment]
Please cherry-pick f39e151659f from [my fork of nixpkgs](https://github.com/matthiasbeyer/nixpkgs).

**Comment by matthiasbeyer** [comment]
Please cherry-pick 4b14f3ebd28 as well.

**Comment by matthiasbeyer** [comment]
Please cherry-pick b019aaa782a as well.

This works for me now. Mycroft cannot understand me, but it at least works.

**Comment by matthiasbeyer** [comment]
Hm, the `HOME` variable setting is still bogus, but it works for me anyways. You propably should have a look at it again.

**Comment by peterhoeg** [comment]
Thanks for trying things out. I had hoped to have it ready for 18.03 but I need to throw some time at it this weekend.

> I pushed another change of the HOME variable thing in 20a7100

The escaping is wrong. Should be ```''$```. Anyway, I'll fix that.

**Comment by matthiasbeyer** [comment]
Any updates on this?

**Comment by peterhoeg** [comment]
Unfortunately not. And I know a lot has changed upstream as well.

**Comment by FRidh** [comment]
Closing due to inactivity.

**Comment by timokau** [comment]
Related: #87621

**Pull request closed without merging**